src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine+AdjustIsolatedMarginInput.kt

```diff
@@ -0,0 +1,146 @@
+            }
+    val childSubaccountNumber = parser.asInt(adjustIsolatedMargin["childSubaccountNumber"])
+    val subaccountNumbers = if (childSubaccountNumber != null) {
+        iListOf(parentSubaccountNumber, childSubaccountNumber)
+    } else {
+        iListOf(parentSubaccountNumber)
+    }
+
+    if (typeText != null) {
+        if (validAdjustIsolatedMarginInput(adjustIsolatedMargin, parentSubaccountNumber, typeText)) {
+            when (typeText) {
+                AdjustIsolatedMarginInputField.Type.rawValue -> {
+                    if (adjustIsolatedMargin["type"] != parser.asString(data)) {
+                        adjustIsolatedMargin.safeSet(typeText, parser.asString(data))
+                        adjustIsolatedMargin.safeSet("amount", null)
+                    }
+                    changes = StateChanges(
+                        iListOf(Changes.wallet, Changes.subaccount, Changes.input),
+                        null,
+                        subaccountNumbers,
+                    )
+                }
+                AdjustIsolatedMarginInputField.Amount.rawValue -> {
+                    val amount = parser.asString(data)
+                    adjustIsolatedMargin.safeSet(typeText, amount)
+                    changes = StateChanges(
+                        iListOf(Changes.wallet, Changes.subaccount, Changes.input),
+                        null,
+                        subaccountNumbers,
+                    )
+                }
+                AdjustIsolatedMarginInputField.ChildSubaccountNumber.rawValue -> {
+                    val childSubaccountNumber = parser.asInt(data)
+                    adjustIsolatedMargin.safeSet(typeText, childSubaccountNumber)
+                    val subaccountNumbers = if (childSubaccountNumber != null) {
+                        iListOf(parentSubaccountNumber, childSubaccountNumber)
+                    } else {
+                        iListOf(parentSubaccountNumber)
+                    }
+                    changes = StateChanges(
+                        iListOf(Changes.wallet, Changes.subaccount, Changes.input),
+                        null,
+                        subaccountNumbers,
+                    )
+                }
+                else -> {}
+            }
+        } else {
+            error = cannotModify(typeText)
+        }
+    } else {
+        changes = StateChanges(
+            iListOf(Changes.wallet, Changes.subaccount, Changes.input),
+            null,
+            subaccountNumbers,
+        )
+    }
+
+    input["adjustIsolatedMargin"] = adjustIsolatedMargin
+    this.input = input
+    changes?.let { update(it) }
+    return StateResponse(state, changes, if (error != null) iListOf(error) else null)
+}
+
+fun TradingStateMachine.validAdjustIsolatedMarginInput(
+    adjustIsolatedMargin: Map<String, Any>,
+    parentSubaccountNumber: Int?,
+    typeText: String?,
+): Boolean {
+    if (typeText == null) return false
```

Comment:
if we're using a guard clause like this, can we just make `typeText` non-null?

Reply:
Yep! I can audit all the fn that rely on `type` and ensure they are non null since `AdjustIsolatedMarginInput` should default to `ADD` if `type` is null.

Reply:
hmm in this specific case `typeText` can be null. Calling the input `AdjustIsolatedMarginInput(null, null)` is valid.